### PR TITLE
Marshal alternative stringify and parse

### PR DESCRIPTION
**Pull request by erights**

### CODE REVIEW

**Comment by erights** [comment]
This is ready for review. It is still Monday somewhere so this still counts as EOD ;)

**Comment by erights** [comment]
@warner, I now have some understanding of #2439 . Enough to avoid that bug in this PR, and to talk about it. I have also preserved a fork of this PR as #2444 as a stable point to demonstrate the bug while this one moves ahead.

@FUDCo, I removed the `'M:'` prefix. PTAL

@michaelfig, @FUDCo's argument to remove `'M:'` that convinced me is interesting and different than I thought. It is a form of the end-to-end argument. Any use of JSON uses some encoding to encode meaning into a JSON structure. The end points of a JSON communication --- the ultimate source and the ultimate destination --- must both coordinate on this meaning for the JSON to be meaningful anyway. It is the tool in the middle, like `jq` and many others, that can be used to manipulate JSON without know how meaning is encoded into that JSON. By removing the `'M:'` we do make it easier for the ultimate endpoints to mis-coordinate. But all uses of JSON have this hazard because all use some idiosyncratic encoding. The benefit of JSON is the genericness of the tools in the middle. Our use of `'@qclass'` falls within this framework. It is not an exception to it.

@FUDCo, did I represent your argument well?

**Comment by FUDCo** [comment]
> @FUDCo, did I represent your argument well?

One aspect of it, yes, very well.  The other aspect is that in some cases you don't know a priori who the other end is going to be, and therefor don't know if they're going to be in a position to use your special software to interpret the message.  But I think the argument based on `jq` and the like is sufficient to settle the matter, so I'm very happy with the outcome.

File: packages/marshal/README.md [hunk 1]
```diff
@@ -128,3 +128,39 @@ Each time `m.unserialize()` encounters such a record, it calls
...
+
+The JSON methods have more than one parameter, enabling customization
+of the operation, for example with *replacers* or *revivers*. These
+marshal-based alternative do not.
+
+The full marshal package will serialize `Passable` objects containing
+presences and promises, because it serializes to a `CapData` structure
+containing both a `body` string and a `slots` array. Marshal's `stringify`
+function serializes only to a string, and so will not
+accept any presences of promises. If any are found in the input, this
```
Comment (FUDCo) [inline packages/marshal/README.md]: ^of^or

**Review by FUDCo — APPROVED** [review]
Other than one comment typo that I flagged, this looks fine.

**Comment by warner** [comment]
Using some distinguisher/prevent-accidental-misinterpretation like `M:` makes sense to me, and makes it clear that the two sides are 1: not speaking JSON (rather some derived/enhanced/augmented thing which looks enough like JSON to fool you without a warning sticker) and 2: really need to be using this particular library (or one explicitly marked as compatible with it) to get the semantics right. That library doesn't necessarily have to be written in JavaScript: cross-language interconnect is super-important sooner or later. But it does suggest that, whatever languages you might be using, both ends of the "wire" are certainly using a library that comes from e.g. `@agoric/something`.

The benefit of using plain JSON is improved generality. An API which is meant to be consumed by something outside your particular ecosystem will benefit from using an encoding that gives you more freedom to choose the decoder implementation. There are still semantics on top of that (what keys do you expect? what should you do with the values?), but most application authors are going to have to write that code anyways. So I imagine they'd prefer a plain JSON encoding plus their hand-written handlers for the semantics, over a vendor-provided decoding library (which maybe needs to be ported to their particular environment) plus the same hand-written handlers for the semantics. Thus, if the application domain doesn't really require datatypes that cannot fit within standard JSON, it's marginally nicer to stick with the standard and reduce the burden on the client developers.

I think of this as an argument for e.g. SwingSet's "mailbox device", which exposes data to the host application (cosmic-swingset), to use plain JSON (specifically using Number rather than BigInt for the sequence numbers). We could expose some `@qclass` to encode BigInt seqnums in a form that could be decoded by our own library, if the other end happens to be using our ecosystem, but it's less proprietary (and not a huge functional loss, 2^53 is plenty big) to stick with Numbers.

**Comment by warner** [comment]
To be clear, I think this could be a useful utility, so please go ahead and land this. I'm just reminding myself/ourselves that the choice between this codec format/library and a plain JSON.stringify should be a deliberate part of the API surface.

**Review by michaelfig — COMMENTED** [review]
Can you please finish the Hilbert Hotel implementation?  It is possible for untrusted JSON to be sucessfully `JSON.parse`d, then serialised with this stringify and throw an exception.

**Comment by erights** [comment]
> Can you please finish the Hilbert Hotel implementation? It is possible for untrusted JSON to be sucessfully `JSON.parse`d, then serialised with this stringify and throw an exception.

Good point! Should I do this as part of this PR before merging? That seems a reasonable requirement. I'll re-ask for your review so you can re-decide when to approve.

On the timing, until the empty object issue is fixed, this PR is not yet that useful anyway.

**Comment by erights** [comment]
Actually the previous approval wasn't from you anyway. Please request revisions if appropriate so I'll see less green ;)

**Review by warner — APPROVED** [review]
Looks good.

In the penultimate paragraph of the README, it might be useful to put a name to the Hilbert Hotel as the fix-not-chosen-at-this-time it describes.

The way one would import these new functions (`import { parse, stringify } from '@agoric/marshal';`) leaves something minor to be desired.. I don't think I'd want a top-level `parse` floating around my module's namespace, it's kind of ambiguous. `JSON.parse` at least tells you how it wants to be fed. We can't do `import marshal from '@agoric/marshal'; marshal.parse(..)` because that's not how default exports work in ESM, and we wouldn't want to do `import { marshal } from '@agoric/marshal'; marshal.parse(..)` because that'd be even more confusing, and besides `marshal` is just as ambiguous as a bare `parse` or `stringify`, so a proper solution would require inventing a new proper noun for the sort of codec this provides. Of course the caller can always `import { parse: agoricParse, stringify: agoricStringify } from '@agoric/marshal';` or something. I can't think of a good solution, and this is only a minor nuisance, so I won't recommend changing anything.

**Comment by warner** [comment]
I need this for #2490, even without the Hilbert Hotel, but I agree that empty objects (#2018) are a blocker. Which is also waiting on me.

**Comment by erights** [comment]
> Can you please finish the Hilbert Hotel implementation? It is possible for untrusted JSON to be sucessfully `JSON.parse`d, then serialised with this stringify and throw an exception.

In progress at #2465

**Comment by erights** [comment]
@michaelfig with the Hilbert Hotel in progress at #2465 and with approvals in hand, are you ok with me merging this?

**Comment by michaelfig** [comment]
> @michaelfig with the Hilbert Hotel in progress at #2465 and with approvals in hand, are you ok with me merging this?

Yes, go ahead!

**Pull request merged**